src/Control/Effect/Cull.hs

```diff
@@ -44,73 +44,42 @@ cull m = send (Cull m pure)
 -- | Run a 'Cull' effect. Branches outside of any 'cull' block will not be pruned.
 --
 --   prop> run (runNonDet (runCull (pure a <|> pure b))) == [a, b]
-runCull :: (Alternative m, Carrier sig m) => CullC m a -> m a
-runCull = (>>= runBranch (const empty)) . runReader False . runCullC
-
-newtype CullC m a = CullC { runCullC :: ReaderC Bool m (Branch m () a) }
-  deriving (Functor)
-
-instance Alternative m => Applicative (CullC m) where
-  pure = CullC . pure . Pure
-  CullC f <*> CullC a = CullC (liftA2 (<*>) f a)
-
-instance (Alternative m, Carrier sig m) => Alternative (CullC m) where
-  empty = CullC (pure (None ()))
-  l <|> r = CullC $ do
-    res <- runCullC l
-    case res of
-      None _ -> runCullC r
-      Pure a -> do
-        cull <- ask
-        if cull then
-          pure res
-        else
-          pure (Alt (pure a) (runReader cull (runCullC r) >>= runBranch (const empty)))
-      Alt _ _ -> pure res
-
-instance (Alternative m, Monad m) => Monad (CullC m) where
-  CullC m >>= f = CullC (m >>= \case
-    None e    -> pure (None e)
-    Pure a    -> runCullC (f a)
-    Alt m1 m2 -> ReaderC (\ cull -> let k = runReader cull . runCullC . f in (m1 >>= k) <|> (m2 >>= k)))
-
-instance (Alternative m, MonadFail m) => MonadFail (CullC m) where
-  fail s = CullC (fail s)
-
-instance (Alternative m, MonadIO m) => MonadIO (CullC m) where
-  liftIO io = CullC (Pure <$> liftIO io)
-
-instance (Alternative m, Carrier sig m) => MonadPlus (CullC m)
-
-instance (Alternative m, Carrier sig m, Effect sig) => Carrier (Cull :+: NonDet :+: sig) (CullC m) where
-  eff (L (Cull m k))     = CullC (local (const True) (runCullC m) >>= bindBranch (runCullC . k))
+runCull :: Alternative m => CullC m a -> m a
+runCull (CullC m) = runNonDetC (runReader False m) ((<|>) . pure) empty
+
+newtype CullC m a = CullC { runCullC :: ReaderC Bool (NonDetC m) a }
+  deriving (Applicative, Functor, Monad, MonadFail, MonadIO)
+
+instance Alternative (CullC m) where
+  empty = CullC empty
+  l <|> r = CullC $ ReaderC $ \ cull -> NonDetC $ \ cons nil -> do
+    runNonDetC (runReader cull (runCullC l))
+      (\ a as -> cons a (if cull then nil else as))
+      (runNonDetC (runReader cull (runCullC r)) cons nil)
```

Comment:
Defining the functionality in the `Alternative` instance means callers don’t need a `Carrier` instance to use the type via the `Alternative` interface, only if they want to use its additional functionality (e.g. `cull`).